File: Assets/Typescript~/src/Server/Services/Global/Player/PlayerService.ts
import { OnStart, Service } from "@easy-games/flamework-core";
import { ServerSignals } from "Server/ServerSignals";
import { Network } from "Shared/Network";
import { Player } from "Shared/Player/Player";
import { Signal } from "Shared/Util/Signal";

@Service({})
export class PlayerService implements OnStart {
	/** Fires when a player first connects to the server. */
	public readonly PlayerPreReady = new Signal<[player: Player]>();

	public readonly PlayerAdded = new Signal<[player: Player]>();

	/** Fires when a player is removed from the game. */
	public readonly PlayerRemoved = new Signal<[player: Player]>();

	private readonly players: Player[] = [];
	private playersPendingReady = new Map<number, Player>();

	constructor() {
		const playerCore = GameObject.Find("Players").GetComponent<PlayerCore>();
		const onPlayerPreJoin = (clientInfo: ClientInfoDto) => {
			const player = new Player(
				clientInfo.GameObject.GetComponent<NetworkObject>(),
				clientInfo.ClientId,
				clientInfo.UserId,
				clientInfo.Username,
				clientInfo.UsernameTag,
			);
			clientInfo.GameObject.name = `Player_${clientInfo.Username}`;
			this.playersPendingReady.set(clientInfo.ClientId, player);
			this.PlayerPreReady.Fire(player);
		};
		const onPlayerRemoved = (clientInfo: ClientInfoDto) => {
			const clientId = clientInfo.ClientId;
			const index = this.players.findIndex((player) => player.clientId === clientId);
			if (index === -1) return;
			const player = this.players[index];
			this.players.remove(index);
			this.PlayerRemoved.Fire(player);
			ServerSignals.PlayerLeave.fire(player);
			Network.ServerToClient.RemovePlayer.Server.FireAllClients(player.clientId);
			player.Destroy();
		};
		const players = playerCore.GetPlayers();
		for (let i = 0; i < players.Length; i++) {
			const clientInfo = players.GetValue(i);
			onPlayerPreJoin(clientInfo);
		}
		playerCore.OnPlayerAdded((clientInfo) => {
			onPlayerPreJoin(clientInfo);
		});
		playerCore.OnPlayerRemoved((clientInfo) => {
			onPlayerRemoved(clientInfo);
		});

		// Player completes join
		Network.ClientToServer.Ready.Server.OnClientEvent((clientId) => {
			if (!this.playersPendingReady.has(clientId)) {
				//print("player not found in pending: " + clientId);
				error("Player not found in pending: " + clientId);
			}

			const player = this.playersPendingReady.get(clientId)!;
			this.playersPendingReady.delete(clientId);
			this.players.push(player);

			Network.ServerToClient.AddPlayer.Server.FireAllClients(player.Encode());
			Network.ServerToClient.AllPlayers.Server.FireClient(
				player.clientId,
				this.players.map((p) => p.Encode()),
			);
			this.PlayerAdded.Fire(player);
			ServerSignals.PlayerJoin.fire(player);
		});
	}

	/** Get all players. */
	public GetPlayers(): Readonly<Array<Player>> {
		return this.players;
	}

	/** Attempt to retrieve a player by `clientId`. */
	public GetPlayerFromClientId(clientId: number): Player | undefined {
		const player = this.players.find((player) => player.clientId === clientId);
		return player;
	}

	/** Attempt to retrieve a player by username. */
	public GetPlayerFromUsername(name: string): Player | undefined {
		const player = this.players.find((player) => player.username === name);
		return player;
	}

	/**
	 * Observe every player entering/leaving the game. The returned function can be
	 * called to stop observing.
	 *
	 * The `observer` function is fired for every player currently in the game and
	 * every future player that joins. The `observer` function must return another
	 * function which is called when said player leaves (_or_ the top-level observer
	 * function was called to stop the observation process).
	 *
	 * ```ts
	 * playersService.ObservePlayers((player) => {
	 * 	print(`${player.name} entered`);
	 * 	return () => {
	 * 		print(`${player.name} left`);
	 * 	};
	 * });
	 * ```
	 */
	public ObservePlayers(observer: (player: Player) => (() => void) | void): () => void {
		const cleanupPerPlayer = new Map<Player, () => void>();
		const observe = (player: Player) => {
			const cleanup = observer(player);
			if (cleanup !== undefined) {
				cleanupPerPlayer.set(player, cleanup);
			}
		};
		for (const player of this.players) {
			observe(player);
		}
		const stopPlayerAdded = this.PlayerAdded.Connect((player) => {
			observe(player);
		});
		const stopPlayerRemoved = this.PlayerRemoved.Connect((player) => {
			const cleanup = cleanupPerPlayer.get(player);
			if (cleanup !== undefined) {
				cleanup();
				cleanupPerPlayer.delete(player);
			}
		});
		return () => {
			stopPlayerAdded();
			stopPlayerRemoved();
			for (const [player, cleanup] of cleanupPerPlayer) {
				cleanup();
			}
		};
	}

	OnStart() {}
}

File: Assets/Bundles/Server/Resources/TS/Services/Global/Player/PlayerService.lua
-- Compiled with unity-ts v2.1.0-75
local Reflect = require("Shared/rbxts_include/node_modules/@easy-games/flamework-core/out/init").Reflect
local Service = require("Shared/rbxts_include/node_modules/@easy-games/flamework-core/out/init").Service
local ServerSignals = require("Server/TS/ServerSignals").ServerSignals
local Network = require("Shared/TS/Network").Network
local Player = require("Shared/TS/Player/Player").Player
local Signal = require("Shared/TS/Util/Signal").Signal
local PlayerService
do
	PlayerService = setmetatable({}, {
		__tostring = function()
			return "PlayerService"
		end,
	})
	PlayerService.__index = PlayerService
	function PlayerService.new(...)
		local self = setmetatable({}, PlayerService)
		return self:constructor(...) or self
	end
	function PlayerService:constructor()
		self.PlayerPreReady = Signal.new()
		self.PlayerAdded = Signal.new()
		self.PlayerRemoved = Signal.new()
		self.players = {}
		self.playersPendingReady = {}
		local playerCore = GameObject:Find("Players"):GetComponent("PlayerCore")
		local onPlayerPreJoin = function(clientInfo)
			local player = Player.new(clientInfo.GameObject:GetComponent("NetworkObject"), clientInfo.ClientId, clientInfo.UserId, clientInfo.Username, clientInfo.UsernameTag)
			clientInfo.GameObject.name = "Player_" .. clientInfo.Username
			local _playersPendingReady = self.playersPendingReady
			local _clientId = clientInfo.ClientId
			_playersPendingReady[_clientId] = player
			self.PlayerPreReady:Fire(player)
		end
		local onPlayerRemoved = function(clientInfo)
			local clientId = clientInfo.ClientId
			local _players = self.players
			local _arg0 = function(player)
				return player.clientId == clientId
			end
			-- ▼ ReadonlyArray.findIndex ▼
			local _result = -1
			for _i, _v in _players do
				if _arg0(_v, _i - 1, _players) == true then
					_result = _i - 1
					break
				end
			end
			-- ▲ ReadonlyArray.findIndex ▲
			local index = _result
			if index == -1 then
				return nil
			end
			local player = self.players[index + 1]
			table.remove(self.players, index + 1)
			self.PlayerRemoved:Fire(player)
			ServerSignals.PlayerLeave:fire(player)
			Network.ServerToClient.RemovePlayer.Server:FireAllClients(player.clientId)
			player:Destroy()
		end
		local players = playerCore:GetPlayers()
		do
			local i = 0
			local _shouldIncrement = false
			while true do
				if _shouldIncrement then
					i += 1
				else
					_shouldIncrement = true
				end
				if not (i < players.Length) then
					break
				end
				local clientInfo = players:GetValue(i)
				onPlayerPreJoin(clientInfo)
			end
		end
		playerCore:OnPlayerAdded(function(clientInfo)
			onPlayerPreJoin(clientInfo)
		end)
		playerCore:OnPlayerRemoved(function(clientInfo)
			onPlayerRemoved(clientInfo)
		end)
		-- Player completes join
		Network.ClientToServer.Ready.Server:OnClientEvent(function(clientId)
			local _playersPendingReady = self.playersPendingReady
			local _clientId = clientId
			if not (_playersPendingReady[_clientId] ~= nil) then
				-- print("player not found in pending: " + clientId);
				error("Player not found in pending: " .. tostring(clientId))
			end
			local _playersPendingReady_1 = self.playersPendingReady
			local _clientId_1 = clientId
			local player = _playersPendingReady_1[_clientId_1]
			local _playersPendingReady_2 = self.playersPendingReady
			local _clientId_2 = clientId
			_playersPendingReady_2[_clientId_2] = nil
			table.insert(self.players, player)
			Network.ServerToClient.AddPlayer.Server:FireAllClients(player:Encode())
			local _fn = Network.ServerToClient.AllPlayers.Server
			local _exp = player.clientId
			local _players = self.players
			local _arg0 = function(p)
				return p:Encode()
			end
			-- ▼ ReadonlyArray.map ▼
			local _newValue = table.create(#_players)
			for _k, _v in _players do
				_newValue[_k] = _arg0(_v, _k - 1, _players)
			end
			-- ▲ ReadonlyArray.map ▲
			_fn:FireClient(_exp, _newValue)
			self.PlayerAdded:Fire(player)
			ServerSignals.PlayerJoin:fire(player)
		end)
	end
	function PlayerService:GetPlayers()
		return self.players
	end
	function PlayerService:GetPlayerFromClientId(clientId)
		local _players = self.players
		local _arg0 = function(player)
			return player.clientId == clientId
		end
		-- ▼ ReadonlyArray.find ▼
		local _result
		for _i, _v in _players do
			if _arg0(_v, _i - 1, _players) == true then
				_result = _v
				break
			end
		end
		-- ▲ ReadonlyArray.find ▲
		local player = _result
		return player
	end
	function PlayerService:GetPlayerFromUsername(name)
		local _players = self.players
		local _arg0 = function(player)
			return player.username == name
		end
		-- ▼ ReadonlyArray.find ▼
		local _result
		for _i, _v in _players do
			if _arg0(_v, _i - 1, _players) == true then
				_result = _v
				break
			end
		end
		-- ▲ ReadonlyArray.find ▲
		local player = _result
		return player
	end
	function PlayerService:ObservePlayers(observer)
		local cleanupPerPlayer = {}
		local observe = function(player)
			local cleanup = observer(player)
			if cleanup ~= nil then
				local _player = player
				cleanupPerPlayer[_player] = cleanup
			end
		end
		for _, player in self.players do
			observe(player)
		end
		local stopPlayerAdded = self.PlayerAdded:Connect(function(player)
			observe(player)
		end)
		local stopPlayerRemoved = self.PlayerRemoved:Connect(function(player)
			local _player = player
			local cleanup = cleanupPerPlayer[_player]
			if cleanup ~= nil then
				cleanup()
				local _player_1 = player
				cleanupPerPlayer[_player_1] = nil
			end
		end)
		return function()
			stopPlayerAdded()
			stopPlayerRemoved()
			for player, cleanup in cleanupPerPlayer do
				cleanup()
			end
		end
	end
	function PlayerService:OnStart()
	end
end
-- (Flamework) PlayerService metadata
Reflect.defineMetadata(PlayerService, "identifier", "Bundles/Server/Services/Global/Player/PlayerService@PlayerService")
Reflect.defineMetadata(PlayerService, "flamework:implements", { "$:flamework@OnStart" })
Reflect.decorate(PlayerService, "$:flamework@Service", Service, { {} })
return {
	PlayerService = PlayerService,
}
-- ----------------------------------
-- ----------------------------------
